### ISSUE TITLE
[Performance]: 3.4 s engine stalls from DeepGEMM compiling the o-projection kernel per prefill chunk size, DeepSeek-V4.1-Flash on 8x B200

### ISSUE THREAD

**Issue by TarzanZhao** [vllm-project/vllm#56684]
### Misc discussion on performance

I ran `vllm bench throughput` on `deepseek-ai/DeepSeek-V4.1-Flash` (`--tensor-parallel-size 8`, ShareGPT, 1000 prompts, seed 42) on 8x B200 with the precompiled-wheel editable install from the CUDA install doc, and got 24.6 s for the 1000 requests in a clean run but 27.6 to 27.9 s in about one run out of three. A per-step timer showed one prefill step taking 3.4 s instead of 0.15 s, always a chunk with a token count the process had not seen before (5908, 7241, 10349, 6109, 2125, 1999 tokens in my runs), and a stack dump of the worker during the stall ended in `deep_gemm.fp8_einsum` called from `vllm/models/deepseek_v4/nvidia/ops/o_proj.py`: DeepGEMM picks the kernel configuration of the fp8 o-projection einsum from the token count and JIT-compiles a new `sm100_fp8_gemm_1d1d` kernel for each configuration it has not met, 3 to 3.6 s of nvcc each, in the middle of serving. With the o-projection input padded to 1024-token buckets above the CUDA graph range and those buckets compiled in the kernel warm-up, the stall did not recur: from an empty DeepGEMM cache, the unmodified code stalled three times in a run and took 26.19 s, the padded one stalled zero times and took 22.95 s (that number also includes the other changes of the PR below; the stall itself is 3.4 s).

Nothing in the logs pointed at it. `jit_monitor` reports Triton and CuTeDSL compiles during inference and stayed quiet, because DeepGEMM's compiles are not in its scope; the only trace was a new `kernel.sm100_fp8_gemm_1d1d.<hash>/kernel.cubin` appearing under `$VLLM_CACHE_ROOT/deep_gemm/cache` at the time of each stall (40 variants of that kernel by the end of my runs). A DeepGEMM compile hook in `jit_monitor`, or a log line from the DeepGEMM wrapper naming the kernel and the shape when it compiles during inference, would have saved me the stack dumps.

### Proposal to improve performance

I opened #56686 to record my experiment: setup, measurements, traces and some potential fixes. A different fix may also well suit the codebase. The PR pads the eager o-projection token count to a bucket and pre-compiles the buckets at warm-up; the indexer's prefill logits kernel (`fp8_fp4_mqa_logits`) still compiles once on the first prefill request of a process with an empty cache, and may want the same treatment.

### Your current environment (if you think it is necessary)

```text
OS: Ubuntu 22.04.5 LTS (x86_64)
GPU: 8x NVIDIA B200, driver 580.126.20, CUDA runtime 13.0
Python 3.12.14, PyTorch 2.13.0+cu132
vLLM main @ 988d9b6777, editable install with the precompiled cu130 wheel (VLLM_USE_PRECOMPILED=1)
flashinfer-python 0.6.18.post1, triton 3.7.1, nvidia-nccl-cu13 2.29.7, transformers 5.17.0, vendored DeepGEMM
```

**Comment by CyFeng16** [comment]
Related observation from an SM90 deployment: same shape, different kernel family.

Host: 4x H200, vLLM 0.29.1rc1.dev12+g410f6da5c, DeepSeek-V4.1-Flash, TP4, `--max-num-seqs 16`, `--max-num-batched-tokens 8192`, up 48 h.

- Our DeepGEMM cache holds 736 kernels, 587 of them `sm90_fp8_gemm_1d2d`. We have no `sm90_fp8_gemm_1d1d`, so your `sm100_fp8_gemm_1d1d` looks like the SM100-family counterpart of our `1d2d`.
- They were written in five batches (2026-05-06, 06-13, 07-26, 08-01, 08-28; 162-188 each, consistent with warm-up), and nothing has been added since 2026-08-28 - i.e. no DeepGEMM compile during serving on this host in 18 days, including the current 48 h run.
- We do get in-serving JIT spikes, but from Triton rather than DeepGEMM: `jit_monitor` logged 51 `Triton kernel JIT compilation during inference` warnings here (`_topp_sb_step_kernel`, `_gumbel_sample_kernel`, `_topk_topp_kernel`, `CombineTopkSwaIndicesKernel.kernel`), most recently 2026-09-15 08:21.

That corroborates the scope gap you describe: `jit_monitor` reports Triton compiles and nothing from DeepGEMM. Our cache happened to be warm; this model has never been run cold on this host, so I cannot say whether SM90 would stall the same way from an empty cache.

Related, possibly useful for #56686: the indexer-logits buffer sized by `max_model_len` is visible on our side of the trade-off too. We run `--max-model-len 1,048,576` but only 16 concurrent sequences, so `[requests, max_model_len]` fp32 works out to 64 MiB per call for us, against 3.5 GB at 830 requests in your report. Same code path, very different bill.
